# What makes resturants close

In this part of the analysis we wish to investigate what factors makes resturants close.

## Load data

In [1]:
import pandas as pd
import numpy as np

In [2]:
Resturants=pd.read_csv('data/toronto_restaurant_business.csv',encoding="utf8")

In [3]:
len(Resturants)

8681

## Data preparation

First we remove data columns which decribe information which is only specific for the individual resturant. This will be the business id, the name of the resturant, address and since we are only intrested in resturants in Toronto also city and state. While longitude and latitude provide more precises information the information is overlapping with the postal_code, however the postal code provide a more general indication of the area and are therefore choosen to represent location.

In [4]:
Resturants

business_id                                    name  \
0     rVBPQdeayMYht4Uv_FOLHg                  Gourmet Burger Company   
1     0QjROMVW9ACKjhSEfHqNCQ                        Mi Mi Restaurant   
2     8k62wYhDVq1-652YbJi5eg                             Tim Hortons   
3     0DnQh8SE8BSnvJltGCCiWg                             Chick-N-Joy   
4     NLaK58WvlNQdUunSIkt-jA               Zav Coffee Shop & Gallery   
...                      ...                                     ...   
8676  thzyiQZb16zD8wMliaEfRQ                           Sushi Supreme   
8677  eUi3O-8Gqh_nJ_ZhO-25gQ                        Get & Go Burrito   
8678  yFQCdWr_k1pTObzHPGis9Q                  Grasshopper Restaurant   
8679  GAgEoHcf4PSuZRS5Zd3ltA                            Q's Shawarma   
8680  n2BKYb8aw4AhRKLkdUjQRg  Steak & Cheese & Quick Pita Restaurant   

                         address     city state postal_code   latitude  \
0             843 Kipling Avenue  Toronto    ON     M8Z 5G9  43.633291   
1           688 Gerrard Street E  Toronto    ON     M4M 1Y3  43.666376   
2           90 Adelaide Street W  Toronto    ON     M5H 3V9  43.649859   
3         3-1265 York Mills Road  Toronto    ON     M3A 1Z3  43.765279   
4           2048 Danforth Avenue  Toronto    ON     M4C 1J6  43.685608   
...                          ...      ...   ...         ...        ...   
8676           1995 Yonge Street  Toronto    ON     M4S 1Z8  43.700617   
8677  1077 Wilson Avenue, Unit 8  Toronto    ON     M3K 1G7  43.726656   
8678          310 College Street  Toronto    ON     M5T 1S2  43.657716   
8679      1075 Martin Grove Road  Toronto    ON     M9W 4W6  43.701807   
8680              199 College St  Toronto    ON     M5T 1P9  43.658683   

      longitude  stars  review_count  is_open  \
0    -79.531768    3.0            13        0   
1    -79.348773    4.0           116        1   
2    -79.382060    3.0             8        1   
3    -79.326248    3.0            11        1   
4    -79.313936    4.5            24        1   
...         ...    ...           ...      ...   
8676 -79.396762    4.0           104        1   
8677 -79.480365    3.5            43        1   
8678 -79.402098    4.0           177        1   
8679 -79.575135    4.0            16        1   
8680 -79.395071    2.5            21        1   

                                             attributes  \
0     {"RestaurantsPriceRange2": "2", "RestaurantsAt...   
1     {"RestaurantsTakeOut": "True", "Alcohol": "bee...   
2     {"OutdoorSeating": "False", "RestaurantsDelive...   
3     {"NoiseLevel": "loud", "BusinessParking": {"ga...   
4     {"DogsAllowed": "False", "OutdoorSeating": "Tr...   
...                                                 ...   
8676  {"Ambience": {"romantic": "False", "intimate":...   
8677  {"BusinessParking": {"garage": "False", "stree...   
8678  {"DogsAllowed": "False", "BikeParking": "True"...   
8679  {"GoodForKids": "True", "HasTV": "True", "Rest...   
8680  {"OutdoorSeating": "False", "Alcohol": "none",...   

                                             categories  \
0                            Restaurants, Burgers, Food   
1                               Vietnamese, Restaurants   
2     Bagels, Donuts, Food, Cafes, Coffee & Tea, Res...   
3                  Fast Food, Restaurants, Chicken Shop   
4           Coffee & Tea, Restaurants, Sandwiches, Food   
...                                                 ...   
8676                  Sushi Bars, Restaurants, Japanese   
8677                               Mexican, Restaurants   
8678              Vegan, Restaurants, Vegetarian, Salad   
8679                         Restaurants, Mediterranean   
8680                            Sandwiches, Restaurants   

                                                  hours  
0                                                    {}  
1     {"Monday": "11:00-22:00", "Tuesday": "11:00-22...  
2                                                    {}  
3                       

In [5]:
Resturants=Resturants.drop(['name','address','city','state','longitude','latitude','business_id'],axis=1)

To be able to work with the data postal_code is one-hot-encoded.

In [6]:
Resturants=pd.concat([Resturants,pd.get_dummies(Resturants.postal_code,prefix='postal_code',drop_first=False)],axis=1)
Resturants=Resturants.drop(['postal_code'],axis=1)

The atributtes ne

In [7]:
# attribute is transformed from string to dictionaries
Resturants.attributes=Resturants.attributes.apply(lambda d: eval(d))

As not all attributes are listed for all resturants we first need to get a list of all available attributes

In [8]:
attributes=set([])
for d in Resturants.attributes:
    attributes=attributes.union(set(d.keys()))
print("The are the following {} attributes:".format(len(attributes)))
for attribute in attributes:
    print("- {}".format(attribute))

The are the following 37 attributes:
- Corkage
- CoatCheck
- RestaurantsTakeOut
- BusinessParking
- BusinessAcceptsBitcoin
- GoodForDancing
- BikeParking
- DietaryRestrictions
- BusinessAcceptsCreditCards
- RestaurantsAttire
- RestaurantsDelivery
- AgesAllowed
- RestaurantsPriceRange2
- RestaurantsCounterService
- WheelchairAccessible
- NoiseLevel
- DogsAllowed
- DriveThru
- Caters
- BYOB
- Ambience
- HappyHour
- OutdoorSeating
- GoodForMeal
- Music
- BestNights
- AcceptsInsurance
- HasTV
- RestaurantsReservations
- ByAppointmentOnly
- RestaurantsGoodForGroups
- HairSpecializesIn
- WiFi
- RestaurantsTableService
- Alcohol
- GoodForKids
- Smoking


Each attribute is now represented as individual columns in the dataframe.

In [11]:
print("Percent of resturants for which attribute is NaN:")
print("---------------------------------------------------------")
for attribute in attributes:
    Resturants[attribute]=Resturants.attributes.apply(lambda a:  a[attribute] if (attribute in a) else np.nan)
    print("{}: {}%".format(attribute,Resturants[attribute].isna().sum()/len(Resturants[attribute].isna())*100))
Resturants=Resturants.drop(['attributes'],axis=1)

Percent of resturants for which attribute is NaN:
---------------------------------------------------------
Corkage: 99.35491302845294%
CoatCheck: 95.63414353185117%
RestaurantsTakeOut: 15.631839649809931%
BusinessParking: 18.78815804630803%
BusinessAcceptsBitcoin: 99.9769611795876%
GoodForDancing: 94.74714894597398%
BikeParking: 39.00472295818454%
DietaryRestrictions: 99.86176707752563%
BusinessAcceptsCreditCards: 96.8667204239143%
RestaurantsAttire: 26.414007602810734%
RestaurantsDelivery: 22.209422877548672%
AgesAllowed: 99.9769611795876%
RestaurantsPriceRange2: 19.32957032599931%
RestaurantsCounterService: 99.98848058979381%
WheelchairAccessible: 82.67480704987905%
NoiseLevel: 34.48911415735515%
DogsAllowed: 87.29409054256422%
DriveThru: 96.49809929731597%
Caters: 45.89333026149061%
BYOB: 99.066927773298%
Ambience: 21.518258265176822%
HappyHour: 87.67423107936874%
OutdoorSeating: 21.276350650846677%
GoodForMeal: 47.909227047575165%
Music: 92.25895634143532%
BestNights: 94.862343048

As we can see that fare from all attributes are listed for all resturants. However NaN values does not provide information about a resturant, and can therefore not be included when training the classifier. Furthermore we see that certain attributes are not listed for most resturants. In such cases it is assumed that a value for NaN cases cannot gennerally be reasonably derived. Attributes where more than 80% of resturants don't have a value listed, are therefore removed. 

In [12]:
disgarded_attributes=set([])
for attribute in attributes:
    percent_NaN=Resturants[attribute].isna().sum()/len(Resturants[attribute].isna())*100
    if percent_NaN>=80:
        Resturants=Resturants.drop([attribute],axis=1)
        disgarded_attributes=disgarded_attributes.union(set([attribute]))
attributes.difference_update(disgarded_attributes)        
print("Attributes disgarded:")
for a in disgarded_attributes:
      print(a) 
print("-----------------------------")
print("There are {} remaing attributes".format(len(attributes)))

Attributes disgarded:
Corkage
CoatCheck
BusinessAcceptsBitcoin
Music
GoodForDancing
BestNights
DietaryRestrictions
BusinessAcceptsCreditCards
AcceptsInsurance
AgesAllowed
RestaurantsCounterService
ByAppointmentOnly
WheelchairAccessible
HairSpecializesIn
DogsAllowed
DriveThru
BYOB
Smoking
HappyHour
-----------------------------
There are 18 remaing attributes


In [13]:
print("Percent of resturants for which attribute is NaN:")
print("---------------------------------------------------------")
for attribute in attributes:
    print("{}: {}%".format(attribute,Resturants[attribute].isna().sum()/len(Resturants[attribute].isna())*100))

Percent of resturants for which attribute is NaN:
---------------------------------------------------------
RestaurantsTakeOut: 15.631839649809931%
BusinessParking: 18.78815804630803%
BikeParking: 39.00472295818454%
RestaurantsAttire: 26.414007602810734%
RestaurantsDelivery: 22.209422877548672%
RestaurantsPriceRange2: 19.32957032599931%
NoiseLevel: 34.48911415735515%
Caters: 45.89333026149061%
Ambience: 21.518258265176822%
OutdoorSeating: 21.276350650846677%
GoodForMeal: 47.909227047575165%
HasTV: 23.77606266559152%
RestaurantsReservations: 19.68667204239143%
RestaurantsGoodForGroups: 20.94228775486695%
WiFi: 33.19894021426103%
RestaurantsTableService: 66.34028337749108%
Alcohol: 29.24778251353531%
GoodForKids: 22.474369312291213%


The above examples shove that there is a large difference between what kind of values are listed for each attribute. This mean that they must all be treated differently.

For the 18 remaining attributes there is still need to find a solution for NaN values.

In [14]:
print("There are {} resturants with no NaN listed for attributes".format(len(Resturants.dropna())))

There are 1666 resturants with no NaN listed for attributes


As we can see droping all resturants with NaN values are not an option as it will limit the size of the dataset to drastic. 

In [15]:
Resturants.Ambience

0       {'romantic': 'False', 'intimate': 'False', 'cl...
1       {'romantic': 'False', 'intimate': 'False', 'cl...
2       {'touristy': 'False', 'hipster': 'False', 'rom...
3       {'touristy': 'False', 'hipster': 'False', 'rom...
4                                                     NaN
                              ...                        
8676    {'romantic': 'False', 'intimate': 'False', 'cl...
8677    {'romantic': 'False', 'intimate': 'False', 'cl...
8678    {'touristy': 'False', 'hipster': 'False', 'rom...
8679    {'romantic': 'False', 'intimate': 'False', 'cl...
8680    {'touristy': 'False', 'hipster': 'False', 'rom...
Name: Ambience, Length: 8681, dtype: object

In [16]:
Resturants.BikeParking

0         NaN
1        True
2       False
3        True
4        True
        ...  
8676     True
8677    False
8678     True
8679    False
8680     True
Name: BikeParking, Length: 8681, dtype: object

In [17]:
Resturants.NoiseLevel

0       average
1       average
2           NaN
3          loud
4           NaN
         ...   
8676    average
8677    average
8678    average
8679        NaN
8680      quiet
Name: NoiseLevel, Length: 8681, dtype: object

As the above examples show the values in the the attributes are very different and a general solution is there for not available.

### Ambience

In [18]:
Resturants.Ambience.apply(lambda r: type(r)).unique()

array([<class 'dict'>, <class 'float'>, <class 'str'>], dtype=object)

We can see that not all values are dictionaries or NaN values.

In [19]:
for a in Resturants.Ambience:
    if type(a)==str:
        print(a)

None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None
None


We see that values such as None are also available. This appear to coorespond well with NaN. We now take a look at the dictionaries.

In [20]:
ambience_keys=set(Resturants.Ambience[0].keys())
Resturants.Ambience.apply(lambda d: set(d.keys())==ambience_keys if type(d)==dict else True)

0        True
1        True
2       False
3       False
4        True
        ...  
8676     True
8677     True
8678    False
8679     True
8680    False
Name: Ambience, Length: 8681, dtype: bool

We see that there not all ambience dictionaries contains the same keys. We therefor have to collect all keys.

In [22]:
ambience_keys=set([])
for d in Resturants.Ambience:
    if type(d)==dict:
        ambience_keys=ambience_keys.union(set(d.keys()))
print("The are the following {} keys used to describe ambience:".format(len(ambience_keys)))
for key in ambience_keys:
    print("- {}".format(key))

The are the following 9 keys used to describe ambience:
- casual
- upscale
- divey
- intimate
- touristy
- hipster
- classy
- romantic
- trendy


Each are now represented by a new column, where NaN and None are represented as False. Here 0 and 1 are used.

In [24]:
for key in ambience_keys:
    Resturants[key]=Resturants.Ambience.apply(lambda a:  int(bool(a[key])) if type(a)==dict and (key in a) else 0)
Resturants=Resturants.drop(['Ambience'],axis=1)

### Wifi

In [25]:
Resturants.WiFi.unique()

array(['no', 'free', nan, 'paid', 'None'], dtype=object)

We see that there is an overlap of categories, with no and None meaning the same thing. The values are therefor both set to 'no'. To address the issue of NaN, this is here assumed to be 'no' aswell.

In [26]:
Resturants.WiFi=Resturants.WiFi.replace("None","no")
Resturants.WiFi=Resturants.WiFi.replace(np.nan,"no")

Wifi cannow be one-hot encoded.

In [27]:
Resturants=pd.concat([Resturants,pd.get_dummies(Resturants.WiFi,prefix='Wifi_',drop_first=False)],axis=1)
Resturants=Resturants.drop(['WiFi'],axis=1)

### Noise level

In [28]:
Resturants.NoiseLevel.unique()

array(['average', nan, 'loud', 'quiet', 'very_loud', 'None'], dtype=object)

We again see that None and nan is overlapping. When nothing have been mentioned we will here assume average noise level

In [29]:
Resturants.NoiseLevel=Resturants.NoiseLevel.replace("None","average")
Resturants.NoiseLevel=Resturants.NoiseLevel.replace(np.nan,"average")

In [30]:
Resturants=pd.concat([Resturants,pd.get_dummies(Resturants.NoiseLevel,prefix='NoiseLevel_',drop_first=False)],axis=1)
Resturants=Resturants.drop(['NoiseLevel'],axis=1)

### Outdoor seating

In [31]:
Resturants.OutdoorSeating.unique()

array(['False', 'True', nan, 'None'], dtype=object)

Again same trend. If nothing is mentioned we are gonna assume False.

In [32]:
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace("None",0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace(np.nan,0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace('False',0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace('True',1)

### Restaurants TakeOut

In [33]:
Resturants.RestaurantsTakeOut.unique()

array(['True', 'False', nan, 'None'], dtype=object)

Same approach is used

In [34]:
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace("None",0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace(np.nan,0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace('False',0)
Resturants.OutdoorSeating=Resturants.OutdoorSeating.replace('True',1)

### Delivery

In [35]:
Resturants.RestaurantsTakeOut.unique()

array(['True', 'False', nan, 'None'], dtype=object)

In [36]:
Resturants.RestaurantsTakeOut=Resturants.RestaurantsTakeOut.replace("None",0)
Resturants.RestaurantsTakeOut=Resturants.RestaurantsTakeOut.replace(np.nan,0)
Resturants.RestaurantsTakeOut=Resturants.RestaurantsTakeOut.replace('False',0)
Resturants.RestaurantsTakeOut=Resturants.RestaurantsTakeOut.replace('True',1)

### Reservations

In [37]:
Resturants.RestaurantsReservations.unique()

array(['False', nan, 'True', 'None'], dtype=object)

In [38]:
Resturants.RestaurantsReservations=Resturants.RestaurantsReservations.replace("None",0)
Resturants.RestaurantsReservations=Resturants.RestaurantsReservations.replace(np.nan,0)
Resturants.RestaurantsReservations=Resturants.RestaurantsReservations.replace('False',0)
Resturants.RestaurantsReservations=Resturants.RestaurantsReservations.replace('True',1)

### Has TV

In [39]:
Resturants.HasTV.unique()

array(['False', 'True', nan, 'None'], dtype=object)

In [40]:
Resturants.HasTV=Resturants.HasTV.replace("None",0)
Resturants.HasTV=Resturants.HasTV.replace(np.nan,0)
Resturants.HasTV=Resturants.HasTV.replace('False',0)
Resturants.HasTV=Resturants.HasTV.replace('True',1)

# Restaurants Price Range

In [41]:
Resturants.RestaurantsPriceRange2.unique()

array(['2', '1', '3', nan, '4', 'None'], dtype=object)

None and NaN is assumed to be the most common price range

In [42]:
Resturants.RestaurantsPriceRange2=Resturants.RestaurantsPriceRange2.replace('None',np.nan)
Resturants.RestaurantsPriceRange2=Resturants.RestaurantsPriceRange2.apply(lambda l: float(l) if type(l)==str else l)

In [43]:
common_priceRange=Resturants.RestaurantsPriceRange2.mode()
Resturants.RestaurantsPriceRange2=Resturants.RestaurantsPriceRange2.replace(np.nan,common_priceRange)

### Table Service

In [44]:
Resturants.RestaurantsTableService.unique()

array([nan, 'True', 'False', 'None'], dtype=object)

When not listed it is assumed that table service is available.

In [45]:
Resturants.RestaurantsTableService=Resturants.RestaurantsTableService.replace("None",1)
Resturants.RestaurantsTableService=Resturants.RestaurantsTableService.replace(np.nan,1)
Resturants.RestaurantsTableService=Resturants.RestaurantsTableService.replace('False',0)
Resturants.RestaurantsTableService=Resturants.RestaurantsTableService.replace('True',1)

### Alcohol

In [46]:
Resturants.Alcohol.unique()

array(['none', 'beer_and_wine', nan, 'full_bar', 'None'], dtype=object)

None and none are here interpreted to mean that aclohol is not served. Nan is assigned the most common value.

In [47]:
common_alcohol_serving=Resturants.Alcohol.mode()
print(common_alcohol_serving)
Resturants.Alcohol=Resturants.Alcohol.replace(np.nan,common_alcohol_serving)
Resturants.Alcohol=Resturants.Alcohol.replace('none','None')

0    full_bar
dtype: object


In [53]:
Resturants=pd.concat([Resturants,pd.get_dummies(Resturants.Alcohol,prefix='Alcohol',drop_first=False)],axis=1)
Resturants=Resturants.drop(['Alcohol','Alcohol_None'],axis=1)

### Business Parking

In [54]:
Resturants.BusinessParking

0       {'garage': 'False', 'street': 'False', 'valida...
1       {'garage': 'False', 'street': 'True', 'validat...
2       {'garage': 'False', 'street': 'False', 'valida...
3       {'garage': 'False', 'street': 'False', 'valida...
4       {'garage': 'False', 'street': 'False', 'valida...
                              ...                        
8676    {'garage': 'False', 'street': 'True', 'validat...
8677    {'garage': 'False', 'street': 'False', 'valida...
8678    {'garage': 'False', 'street': 'True', 'validat...
8679                                                  NaN
8680    {'garage': 'False', 'street': 'True', 'validat...
Name: BusinessParking, Length: 8681, dtype: object

In [56]:
parking_keys=set([])
for d in Resturants.BusinessParking:
    if type(d)==dict:
        parking_keys=parking_keys.union(set(d.keys()))
print("The are the following {} keys used to discribe parking:".format(len(parking_keys)))
for key in parking_keys:
    print("- {}".format(key))

The are the following 5 keys used to discribe parking:
- valet
- validated
- lot
- garage
- street


In [57]:
for key in parking_keys:
    Resturants[key]=Resturants.BusinessParking.apply(lambda a:  int(bool(a[key])) if type(a)==dict and (key in a) else 0)
Resturants=Resturants.drop(['BusinessParking'],axis=1)

### Catering

In [58]:
Resturants.Caters.unique()

array(['False', nan, 'True', 'None'], dtype=object)

When None or NaN it is assumed that the resturant does not cater

In [59]:
Resturants.Caters=Resturants.Caters.replace("None",0)
Resturants.Caters=Resturants.Caters.replace(np.nan,0)
Resturants.Caters=Resturants.Caters.replace('False',0)
Resturants.Caters=Resturants.Caters.replace('True',1)

### Kid friendly

In [60]:
Resturants.GoodForKids.unique()

array(['True', nan, 'False', 'None'], dtype=object)

In this case it is assumed that if None or NaN then the resturant is not kid friendly.<br>
**Would it make more sense to allow for none, as in not partically kid friendly or bad for kids..?**

In [61]:
Resturants.GoodForKids=Resturants.GoodForKids.replace("None",0)
Resturants.GoodForKids=Resturants.GoodForKids.replace(np.nan,0)
Resturants.GoodForKids=Resturants.GoodForKids.replace('False',0)
Resturants.GoodForKids=Resturants.GoodForKids.replace('True',1)

### Attire

In [62]:
Resturants.RestaurantsAttire.unique()

array(['casual', nan, 'dressy', 'formal', 'None'], dtype=object)

When None available casual is assumed, here we allow for resturants to not have a special attrie.

In [63]:
Resturants.GoodForKids=Resturants.RestaurantsAttire.replace(np.nan,'None')
Resturants=pd.concat([Resturants,pd.get_dummies(Resturants.RestaurantsAttire,prefix='Attire',drop_first=False)],axis=1)
Resturants=Resturants.drop(['RestaurantsAttire','Attire_None'],axis=1)

### Bike parking

In [64]:
Resturants.BikeParking.unique()

array([nan, 'True', 'False', 'None'], dtype=object)

Here NaN and None is interpreted to mean false. **-reasonable for NaN?**

In [65]:
Resturants.BikeParking=Resturants.BikeParking.replace("None",0)
Resturants.BikeParking=Resturants.BikeParking.replace(np.nan,0)
Resturants.BikeParking=Resturants.BikeParking.replace('False',0)
Resturants.BikeParking=Resturants.BikeParking.replace('True',1)

### Suitable for groups

In [66]:
Resturants.RestaurantsGoodForGroups.unique()

array(['True', nan, 'False', 'None'], dtype=object)

In [67]:
Resturants.RestaurantsGoodForGroups=Resturants.RestaurantsGoodForGroups.replace("None",0)
Resturants.RestaurantsGoodForGroups=Resturants.RestaurantsGoodForGroups.replace(np.nan,0)
Resturants.RestaurantsGoodForGroups=Resturants.RestaurantsGoodForGroups.replace('False',0)
Resturants.RestaurantsGoodForGroups=Resturants.RestaurantsGoodForGroups.replace('True',1)

### What type of meals are the resturant good for

In [69]:
Resturants.GoodForMeal

0                                                     NaN
1       {'dessert': 'False', 'latenight': 'False', 'lu...
2                                                     NaN
3       {'dessert': 'False', 'latenight': 'False', 'lu...
4                                                     NaN
                              ...                        
8676    {'dessert': 'False', 'latenight': 'False', 'lu...
8677    {'dessert': 'False', 'latenight': 'False', 'lu...
8678    {'dessert': 'False', 'latenight': 'False', 'lu...
8679                                                  NaN
8680                                                  NaN
Name: GoodForMeal, Length: 8681, dtype: object

In [70]:
meal_keys=set([])
for d in Resturants.GoodForMeal:
    if type(d)==dict:
        meal_keys=meal_keys.union(set(d.keys()))
print("The are the following {} keys used to discribe meals:".format(len(meal_keys)))
for key in meal_keys:
    print("- {}".format(key))

The are the following 6 keys used to discribe meals:
- lunch
- breakfast
- latenight
- dessert
- brunch
- dinner


In [71]:
for key in meal_keys:
    Resturants[key]=Resturants.GoodForMeal.apply(lambda a:  int(bool(a[key])) if type(a)==dict and (key in a) else 0)
Resturants=Resturants.drop(['GoodForMeal'],axis=1)

**Is it still a reasonable approach when GoodForMeal: 47.909227047575165% NaN values?**

## Defining training, test and validation sets

## Predicting if a business close

### Baseline

Random classification as a baseline

To predict which resturants close, we build a binary classifyer classifying wheter or not a business is closed or open. In other words we are intrested in building a binary classify. To do so logistic regression is used.

### Logistic regression

In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
clf = LogisticRegression

In [ ]:
clf = LogisticRegression

## Interpreting factors influencing prediction

In [ ]:
 from sklearn.datasets import load_iris